README.md
dse_studio Cookbook
===================

This cookbook installs and configures  Datastax Studio. 

## Requirements

### Cookbooks 

#### Recommended

- `java` - a jvm needs to be on the machine for `studio`
- `nginx-proxy` - to easily stand up an nginx ssh reverse proxy for studio. 

### DSE Credentials file

Before being able to run `test-kitchen` tests, you need to create a file `dse_creds.yaml`:

```yaml
repo_user: <your dse username>
repo_pass: <your dse password>
```

This file is read from `.kitchen.yml` during integration tests.

Attributes
----------

 * `node['dse']['studio']['version']` = Version to install ('1.0.1').
 * `node['dse']['studio']['download_url']` = Pattern to build the download URL. ("http://#{node['dse']['studio']['repo_user']}:#{node['dse']['studio']['repo_pass']}@downloads.datastax.com/enterprise/datastax-studio-#{node['dse']['studio']['version']}.tar.gz")
 * `node['dse']['studio']['download_to']` = Where to download to ('/tmp')
 * `node['dse']['studio']['user']` = Name of user to create for running studio ('dse').
 * `node['dse']['studio']['group']` = Name of a group to add the user to ('dse').
 * `node['dse']['studio']['install_dir']` = Where to install ("/home/#{node['dse']['studio']['user']}").
 * `node['dse']['studio']['repo_user']` = Datastax account's username. Will cause error if not set (nil).
 * `node['dse']['studio']['repo_pass']` = Datastax account's password. Will cause error if not set (nil).

 * `node['dse']['studio']['service_template']` = 'dse_studio.service.erb'
 * `node['dse']['studio']['configuration_template']` = 'configuration.yaml.erb'
 * `node['dse']['studio']['template_cookbook']` = 'dse_studio'
 * `node['dse']['studio']['startup_params']` = Additional startup parameters for studio ('').

 * `node['dse']['studio']['result_set_size_limit']` = 1000
 * `node['dse']['studio']['max_result_size_bytes']` = 10485760
 * `node['dse']['studio']['execution_timeout_ms']` = 0
 * `node['dse']['studio']['schema_refresh_interval_ms']` = 3000
 * `node['dse']['studio']['traversal_sources']['real_time_traversal_source']` = 'g'
 * `node['dse']['studio']['traversal_sources']['analytic_traversal_source']` = 'a'
 * `node['dse']['studio']['http_port']` = 9091
 * `node['dse']['studio']['http_bind_address']` = 'localhost'
 * `node['dse']['studio']['logging']['file_name']` = 'studio.log'
 * `node['dse']['studio']['logging']['max_file_size']` = '250 MB'
 * `node['dse']['studio']['logging']['max_files']` = 10
 * `node['dse']['studio']['logging']['directory']` = './logs'
 * `node['dse']['studio']['security']['encryption_password_file']` = 'conf/security/security.properties'

 * `node['dse']['studio']['user_data']['base_directory']` = 'null'
 * `node['dse']['studio']['user_data']['connections_directory']` = 'connections'
 * `node['dse']['studio']['user_data']['notebooks_directory']` = 'notebooks'

 * `node['dse']['studio']['connection_management']['idle_timeout_in_seconds']` = 3600
 * `node['dse']['studio']['connection_management']['ide']['connection']['service_port']` = 9091
 * `node['dse']['studio']['connection_management']['ide']['connection']['cache_idle_time_in_seconds']` = 3600
 * `node['dse']['studio']['connection_management']['ide']['connection']['config_monitor_frequency_in_seconds']` = 120
 * `node['dse']['studio']['connection_management']['ide']['file']['service_port']` = 9091
 * `node['dse']['studio']['connection_management']['ide']['file']['contents_json_field']` = 'code'
 * `node['dse']['studio']['connection_management']['ide']['graph_schema_chache_idle_time_in_seconds']` = 600
 
## Usage

### dse_studio::default

```yaml
run_list:
      - recipe[java]
      - recipe[dse_studio::studio]
      - recipe[nginx]
      - recipe[dse_studio::reverse_proxy]
    attributes:
      java:
        install_flavor: 'oracle'
        jdk_version: 8
        oracle:
          accept_oracle_download_terms: true
      dse:
        studio:
          repo_user: ...
          repo_pass: ...
      nginx:
        default_site_enabled: false
```

### Setting up an Nginx Proxy

If you want SSL-encrypted connections, you'll need to front this with a reverse proxy. Here's an example using [nginx-proxy](https://github.com/3ofcoins/chef-cookbook-nginx-proxy/), taken from a packer definition file:

```json
{
  "type": "chef-solo",
  "install_command": "curl -L https://www.chef.io/chef/install.sh | {{if .Sudo}}sudo{{end}} bash -s -- -v {{user `chef-version`}}",
  "cookbook_paths": [ "cookbooks" ],
  "run_list": [
    "java",
    "nginx",
    "dse_studio"
  ],
  "json": {
    "java": {
      "jdk_version": "8",
      "install_flavor": "oracle",
      "oracle": {
        "accept_oracle_download_terms": true
      },
      "alternatives_priority": 20000
    },
    "dse": {
      "studio": {
        "repo_user": "...",
        "repo_pass": "..."
      }
    },
    "nginx": {
      "default_site_enabled": false
    },
    "nginx_proxy": {
      "proxies": {
        "localhost": {
          "url": "http://localhost:9091",
          "ssl_key_path": "/opt/ssl/cert.key",
          "ssl_certificate_path": "/opt/ssl/cert.crt"
        }
      }
    }
  }
}
```

Contributing
------------

The usual. Also as usual, highly welcome! 

 1. Fork the repository on Github
 2. Create a named feature branch (like `add_component_x`)
 3. Write your change
 4. Write tests for your change (if applicable)
 5. Run the tests, ensuring they all pass
 6. Submit a Pull Request using Github

License and Authors
-------------------
Authors: Thorsten Krüger


CHANGELOG.md
dse_studio CHANGELOG
====================

This file is used to list changes made in each version of the dse_studio cookbook.

0.1.1 
-----

- [thor] - Cookbook dependency updates
- [thor] - Updated test kitchen gem to ensure compatibility with ruby 2.4
- [thor] - Cookbook no longer depends on nginx

0.1.0
-----
- [thor] - Initial release of dse_studio


spec/dse_studio_spec.rb
require 'spec_helper'

describe 'dse_studio::studio' do
  let(:chef_run) do
    ChefSpec::SoloRunner.new(platform: 'centos', version: '7.2.1511') do |node|
      node.override['dse']['studio']['version'] = '2.0.0'
      node.override['dse']['studio']['repo_user'] = 'some_user'
      node.override['dse']['studio']['repo_pass'] = 'some_password'
    end.converge(described_recipe)
   end

  it 'downloads a tarfile' do
    expect(chef_run).to extract_tar_extract('http://some_user:some_password@downloads.datastax.com/enterprise/datastax-studio-2.0.0.tar.gz')
  end

  it 'creates user and group' do
    expect(chef_run).to create_user('dse')
    expect(chef_run).to create_group('dse').with(members: ['dse'])
  end

  it 'configures the service' do
    expect(chef_run).to enable_service('dse_studio')
    expect(chef_run).to start_service('dse_studio')
  end

  it 'creates dse studio service' do
    expect(chef_run).to create_template('/etc/systemd/system/dse_studio.service')
  end
end


test/integration/studio/inspec/dse_studio_spec.rb
control 'check dse studio initialization' do

  describe service('dse_studio') do
    it { should be_enabled }
    it { should be_running }
  end

end


recipes/studio.rb
#
# Installs DSE Studio from datastax download page
#

repo_user = node['dse']['studio']['repo_user']
repo_pass = node['dse']['studio']['repo_pass']

raise "need to set node['dse']['studio']['repo_user']" unless repo_user
raise "need to set node['dse']['studio']['repo_pass']" unless repo_pass

install_dir = "#{node['dse']['studio']['install_dir']}/datastax-studio-#{node['dse']['studio']['version']}"

#
# Prepare studio user
#

user node['dse']['studio']['user'] do
  comment "DSE Studio user"
  system true
end

group node['dse']['studio']['group'] do
  action :create
  members node['dse']['studio']['user']
  append true
end

directory node['dse']['studio']['install_dir'] do
  action :create
  recursive true
  owner node['dse']['studio']['user']
  group node['dse']['studio']['group']
end

#
# Install Studio from Tarball
#

tar_extract node['dse']['studio']['download_url'] do
  download_dir '/tmp/'
  target_dir node['dse']['studio']['install_dir']
  creates "#{node['dse']['studio']['install_dir']}/datastax-studio-#{node['dse']['studio']['version']}/bin/server.sh"
  user node['dse']['studio']['user']
end

#
# Configure
#

template "#{install_dir}/conf/configuration.yaml" do
  source node['dse']['studio']['configuration_template']
  cookbook node['dse']['studio']['template_cookbook']
  user node['dse']['studio']['user']
  group node['dse']['studio']['group']
  action :create
  notifies :restart, 'service[dse_studio]'
end

#
# Install Service dse_studio
#

template '/etc/systemd/system/dse_studio.service' do
  source node['dse']['studio']['service_template']
  cookbook node['dse']['studio']['template_cookbook']
  action :create
  notifies :restart, 'service[dse_studio]'
end

service 'dse_studio' do
  action [ :enable, :start ]
end


attributes/default.rb


default['dse']['studio']['version'] = '2.0.0'
default['dse']['studio']['download_url'] = "http://#{node['dse']['studio']['repo_user']}:#{node['dse']['studio']['repo_pass']}@downloads.datastax.com/enterprise/datastax-studio-#{node['dse']['studio']['version']}.tar.gz"
default['dse']['studio']['download_to'] = '/tmp'
default['dse']['studio']['user'] = 'dse'
default['dse']['studio']['group'] = 'dse'
default['dse']['studio']['install_dir'] = "/home/#{node['dse']['studio']['user']}"
default['dse']['studio']['repo_user'] = nil
default['dse']['studio']['repo_pass'] = nil

default['dse']['studio']['service_template'] = 'dse_studio.service.erb'
default['dse']['studio']['configuration_template'] = 'configuration.yaml.erb'
default['dse']['studio']['template_cookbook'] = 'dse_studio'
default['dse']['studio']['startup_params'] = ''

default['dse']['studio']['result_set_size_limit'] = 1000
default['dse']['studio']['max_result_size_bytes'] = 10485760
default['dse']['studio']['execution_timeout_ms'] = 0
default['dse']['studio']['schema_refresh_interval_ms'] = 3000
default['dse']['studio']['traversal_sources']['real_time_traversal_source'] = 'g'
default['dse']['studio']['traversal_sources']['analytic_traversal_source'] = 'a'
default['dse']['studio']['http_port'] = 9091
default['dse']['studio']['http_bind_address'] = 'localhost'
default['dse']['studio']['logging']['file_name'] = 'studio.log'
default['dse']['studio']['logging']['max_file_size'] = '250 MB'
default['dse']['studio']['logging']['max_files'] = 10
default['dse']['studio']['logging']['directory'] = './logs'
default['dse']['studio']['security']['encryption_password_file'] = 'conf/security/security.properties'

default['dse']['studio']['user_data']['base_directory'] = 'null'
default['dse']['studio']['user_data']['connections_directory'] = 'connections'
default['dse']['studio']['user_data']['notebooks_directory'] = 'notebooks'

default['dse']['studio']['connection_management']['idle_timeout_in_seconds'] = 3600
default['dse']['studio']['connection_management']['ide']['connection']['service_port'] = 9091
default['dse']['studio']['connection_management']['ide']['connection']['cache_idle_time_in_seconds'] = 3600
default['dse']['studio']['connection_management']['ide']['connection']['config_monitor_frequency_in_seconds'] = 120
default['dse']['studio']['connection_management']['ide']['file']['service_port'] = 9091
default['dse']['studio']['connection_management']['ide']['file']['contents_json_field'] = 'code'
default['dse']['studio']['connection_management']['ide']['graph_schema_chache_idle_time_in_seconds'] = 600
